lib/block_view_validator.go

```diff
@@ -0,0 +1,2133 @@
+	if err != nil {
+		return nil, errors.Wrapf(err, "UtxoView.GetTopActiveValidatorsByStake: error retrieving entries from db: ")
+	}
+	// Cache top N active ValidatorEntries from the db in the UtxoView.
+	for _, validatorEntry := range dbValidatorEntries {
+		// We only pull ValidatorEntries from the db that are not present in the
+		// UtxoView. As a sanity check, we double-check that the ValidatorEntry
+		// is not already in the UtxoView here.
+		if _, exists := bav.ValidatorMapKeyToValidatorEntry[validatorEntry.ToMapKey()]; !exists {
+			bav._setValidatorEntryMappings(validatorEntry)
+		}
+	}
+	// Pull !isDeleted, active ValidatorEntries from the UtxoView.
+	var validatorEntries []*ValidatorEntry
+	for _, validatorEntry := range bav.ValidatorMapKeyToValidatorEntry {
+		if !validatorEntry.isDeleted && validatorEntry.Status() == ValidatorStatusActive {
+			validatorEntries = append(validatorEntries, validatorEntry)
+		}
+	}
+	// Sort the ValidatorEntries DESC by TotalStakeAmountNanos.
+	sort.Slice(validatorEntries, func(ii, jj int) bool {
+		return validatorEntries[ii].TotalStakeAmountNanos.Cmp(validatorEntries[jj].TotalStakeAmountNanos) > 0
+	})
+	// Return top N.
+	upperBound := int(math.Min(float64(limit), float64(len(validatorEntries))))
+	return validatorEntries[0:upperBound], nil
+}
+
+func (bav *UtxoView) GetGlobalStakeAmountNanos() (*uint256.Int, error) {
+	var globalStakeAmountNanos *uint256.Int
+	var err error
+	// Read the GlobalStakeAmountNanos from the UtxoView.
+	if bav.GlobalStakeAmountNanos != nil {
+		globalStakeAmountNanos = bav.GlobalStakeAmountNanos.Clone()
+	}
+	// If not set, read the GlobalStakeAmountNanos from the db.
+	// TODO: Confirm if the GlobalStakeAmountNanos.IsZero() that we should look in the db.
+	if globalStakeAmountNanos == nil || globalStakeAmountNanos.IsZero() {
+		globalStakeAmountNanos, err = DBGetGlobalStakeAmountNanos(bav.Handle, bav.Snapshot)
+		if err != nil {
+			return nil, errors.Wrapf(err, "UtxoView.GetGlobalStakeAmountNanos: ")
+		}
+		if globalStakeAmountNanos == nil {
+			globalStakeAmountNanos = uint256.NewInt()
+		}
+		// Cache the GlobaleStakeAmountNanos from the db in the UtxoView.
+		bav._setGlobalStakeAmountNanos(globalStakeAmountNanos)
+	}
+	return globalStakeAmountNanos, nil
+}
+
+func (bav *UtxoView) _setValidatorEntryMappings(validatorEntry *ValidatorEntry) {
+	// This function shouldn't be called with nil.
+	if validatorEntry == nil {
+		glog.Errorf("_setValidatorEntryMappings: called with nil entry, this should never happen")
+		return
+	}
+	bav.ValidatorMapKeyToValidatorEntry[validatorEntry.ToMapKey()] = validatorEntry
+}
+
+func (bav *UtxoView) _deleteValidatorEntryMappings(validatorEntry *ValidatorEntry) {
+	// This function shouldn't be called with nil.
+	if validatorEntry == nil {
+		glog.Errorf("_deleteValidatorEntryMappings: called with nil entry, this should never happen")
+		return
+	}
+	// Create a tombstone entry.
+	tombstoneEntry := *validatorEntry
```

Comment:
Should we use .Copy() for all of these to be safe?

Reply:
It looks like in other _delete funcs, we don't copy when creating a tombstone entry. Leaving as is for now. We do de-reference with the `*`.